{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b424985b",
   "metadata": {},
   "source": [
    "<img src= 'img/tree.png' style=\"float:left\"> </img>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2dc1b333",
   "metadata": {},
   "source": [
    "# Process Step by Step"
   ]
  },
  {
   "cell_type": "raw",
   "id": "85690d8c",
   "metadata": {},
   "source": [
    "# Fiest part : create correspondence file for each dataset\n",
    "\n",
    "# 1. Set working directory\n",
    "#setwd(\"/home/dm20/Desktop/Recent_work/Bioinformatics/Scleroderma/Diseases_folder\")\n",
    "\n",
    "# 2. Load required libraries \n",
    "library(RCurl)\n",
    "library(GEOquery)\n",
    "library(limma)\n",
    "library(topGO)\n",
    "library(genefilter)\n",
    "\n",
    "# Perform 3.1 for micro array data\n",
    "# perform 3.2 for RNAseq data\n",
    "\n",
    "# 3.1 For micro-array data perform below steps\n",
    "#gse <- getGEO(filename = \"GSE104174_series_matrix.txt.gz\",destdir=getwd())\n",
    "gse <- read.csv('GSE98793mi_array.csv')\n",
    "d <- factor(c(rep('CTRL', 7),rep('SSc',10)))\n",
    "\n",
    "\n",
    "mod <- model.matrix(~0+d)\n",
    "\n",
    "# print(mod)\n",
    "# dSSc dCTRL dCVCTRL  \n",
    "# 1    0     0       0     1\n",
    "# 2    0     0       1     0\n",
    "\n",
    "fit_1 <- lmFit(gse, mod)\n",
    "# #print(fit_1)\n",
    "contr <- makeContrasts(dCTRL-dT2D,levels = mod)\n",
    "#contr <- makeContrasts(dCD-dCTRL, levels = mod)\n",
    "fit_2 <- contrasts.fit(fit = fit_1, contrasts = contr)\n",
    "fit_3 <- eBayes(fit_2)\n",
    "\n",
    "\n",
    "\n",
    "# \n",
    "# # colnames(fit_3)\n",
    "# #[1] \"dSSc - dCTRL\"   \"dTREAT - dSSc\"  \"dTREAT - dCTRL\"\n",
    "# #print(fit_3)\n",
    "# #show(fit_3)\n",
    "# #n=length(fit_3)\n",
    "# #print(n)\n",
    "# #topTable(fit_3, coef=1, adjust=\"BH\") #coef=1 means dSSc-dCTRL\n",
    "# #topTable(fit_3, coef=2, adjust=\"BH\") #coef=1 means dTRAET-dSSc\n",
    "# \n",
    "#table_result <- topTable(fit_3, coef=1,p.value=1e-2, sort.by = \"logFC\")\n",
    "#table_result <- topTable(fit_3, coef=1,p.value=1e-2, sort.by = \"logFC\")\n",
    "#table_result <- topTable(fit_3, coef=1, n=Inf,p.value=5e-2, sort.by = \"logFC\")\n",
    "#dim(table_result)\n",
    "#[1] 156  32\n",
    "table_result <- topTable(fit_3, coef=1,n=Inf,adjust=\"BH\", sort.by = \"Log_FoldChange\")\n",
    "#> dim(table_result)\n",
    "#[1] 18981    32\n",
    "#[1] 156  32\n",
    "#dim(table_result)\n",
    "# # #[1]  9 11\n",
    "\n",
    "\n",
    "subtable_result <- subset(table_result, select=c(\"ID\",\"Gene.Symbol\",\"adj.P.Val\",\"P.Value\",\"t\",\"B\",\"Log_FoldChange\"))\n",
    "subtable_result <- subtable_result[is.na(subtable_result$ID)==0,]\n",
    "# End of micro-array data part\n",
    "\n",
    "# 3.2 For RNAseq : Start from here          #***************************#\n",
    "# data from Biojupies / Grein\n",
    "subtable_result <- read.csv(\"GSE53239_Bipolar_disorder.csv\", header = TRUE)\n",
    "\n",
    "\n",
    "geneList <- subtable_result$Log_FoldChange\n",
    "#names(geneList) <- subtable_result$Gene.Symbol\n",
    "names(geneList) <- subtable_result$Gene_symbol\n",
    "\n",
    "# 4. Save subtable\n",
    "write.csv(subtable_result,\"GSE78936_table.csv\")          #**********          \n",
    "# nrow(subtable_result[subtable_result$P.Value<0.05,])\n",
    "\n",
    "# 5. Apply condition of LogFC threshold \n",
    "topDiffGenes <- function(allScore){\n",
    "  return(abs(allScore)>1.00)\n",
    "}\n",
    "\n",
    "CD_GOdata <- new(\"topGOdata\",\n",
    "                 description = \"cd study\",\n",
    "                 ontology = \"BP\", \n",
    "                 allGenes = geneList,\n",
    "                 geneSel = topDiffGenes,\n",
    "                 annot = annFUN.org, \n",
    "                 ID = \"symbol\", \n",
    "                 mapping = \"org.Hs.eg.db\",\n",
    "                 nodeSize = 6)\n",
    "\n",
    "# 7. Create genes and GO terms\n",
    "n_sg <- sum(topDiffGenes(geneList))\n",
    "\n",
    "sg <- sigGenes(CD_GOdata)\n",
    "\n",
    "ug <- usedGO(CD_GOdata)\n",
    "\n",
    "# 8. Fisher's Exact Test (and Kolmogorov-Smirnov for evaluation, not described)\n",
    "resultFisher <- runTest(CD_GOdata, algorithm = \"classic\", statistic = \"fisher\")\n",
    "resultKS <- runTest(CD_GOdata, algorithm = \"classic\", statistic = \"ks\")\n",
    "\n",
    "\n",
    "# 9. GO terms tree\n",
    "allRes <- GenTable(CD_GOdata, classic = resultFisher, KS = resultKS, orderBy = \"classic\", topNodes = 30)\n",
    "showSigOfNodes(CD_GOdata,score(resultFisher),firstSigNodes = 5, useInfo = \"all\")\n",
    "printGraph(CD_GOdata,resultFisher,firstSigNodes = 5, fn.prefix = \"SCZ9_GSE78936\", useInfo = \"all\", pdfSW = TRUE) #******\n",
    "\n",
    "# 10. Create text file for the correspondence GO terms - genes (this file is mandatory for the 2nd script)\n",
    "terms <- allRes$GO.ID\n",
    "genes <- genesInTerm(CD_GOdata,terms)\n",
    "for (i in 1:length(terms))\n",
    "{\n",
    "  term <- terms[i]\n",
    "  genes_term <- genes[term][[1]]\n",
    "  # find the genes that are in the list of genes of interest\n",
    "  fact <- genes_term %in% sg\n",
    "  genes_term_2 <- genes_term[fact == TRUE]\n",
    "  genes_term_2 <- paste(genes_term_2, collapse=',')\n",
    "  cat(term,\"genes:\",genes_term_2,\"\\n\", append = TRUE, file = \"SCZ9_GSE78936_correspondence.txt\" )  #******\n",
    "}\n",
    "write.csv(allRes,\"GSE78936_TopGeo.csv\") #********\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
